README.rst
============
 barcodecop
============

Enforce barcode match stringency and read quality in demultiplexed MiSeq reads.

.. image:: https://github.com/nhoffman/barcodecop/actions/workflows/test.yml/badge.svg
    :target: https://github.com/nhoffman/barcodecop/actions/workflows/test.yml

Barcode mis-assignment represents a significant problem for ultra
sensitive assays that strongly interpret the presence of very low
prevalence reads within a specimen. In addition, mis-assignment of
reads between specimens can create the appearance of template
contamination in negative controls.

It turns out that barcode read quality is a predictor of barcode
mis-assignment. Wright and Vetsigian (2016)
(https://dx.doi.org/10.1186%2Fs12864-016-3217-x) showed that an
average barcode quality score threshold of 26 prevented most read
mis-assignment on the Illumina platform.

In addition, the onboard software used for demultiplexing on the
Illumina MiSeq cannot be configured to enforce exact barcode
matches. As a result, a minority of reads (up to about 5% in my tests)
are assigned to a specimen on the basis of a partial barcode
match. Somewhat anecdotally, these less-than-exact matches appear to
have a higher likelihood of mis-assignment as well, but the effect is
dependent on barcode sequence and which combinations of barcodes are
used in the same library.

This package provides the ``barcodecop`` command that uses the index
reads to determine the most prevalent barcode sequence, and removes
reads without exact barcode matches from an accompanying fastq
file. It also filters reads based on average barcode quality score.

Command line arguments::

  usage: barcodecop [-h] [-o OUTFILE] [--snifflimit N] [--head N] [--invert] [-q] [-V] (-c | -f file.fastq[.bz2|.gz]) [--match-filter]
                    [--min-pct-assignment PERCENT] [--strict] [-b FILE] [-C FILE] [--allow-empty] [--qual-filter] [-p MIN_QUAL] [--encoding {phred}]
                    file.fastq[.bz2|.gz] [file.fastq[.bz2|.gz] ...]

  Filter fastq files by enforcing exact barcode match and quality score.

  Input and output files may be compressed as indicated by a .bz2 or .gz
  suffix.

  positional arguments:
    file.fastq[.bz2|.gz]  one or two files containing index reads in fastq format

  options:
    -h, --help            show this help message and exit
    -o OUTFILE, --outfile OUTFILE
                          output fastq
    --snifflimit N        read no more than N records from the index file [10000]
    --head N              limit the output file to N records
    --invert              include only sequences failing filtering criteria
    -q, --quiet           minimize messages to stderr
    -V, --version         Print the version number and exit
    -c, --show-counts     tabulate barcode counts and exit
    -f file.fastq[.bz2|.gz], --fastq file.fastq[.bz2|.gz]
                          reads to filter in fastq format

  Barcode matching options:
    --match-filter        filter reads based on exact match to most common barcode [default: no match filter]
    --min-pct-assignment PERCENT
                          warn (or fail with an error; see --strict) if the most common barcode represents less than PERCENT of the total [90.0]
    --strict              fail if conditions of --min-pct-assignment are not met
    -b FILE, --barcode-counts FILE
                          tabulate barcode counts and store as a CSV
    -C FILE, --read-counts FILE
                          tabulate read counts and store as a CSV
    --allow-empty         accept fastq files which contain zero length seqs/qual strings[default: do not allow empty seqs in fastq files]

  Barcode quality filtering options:
    --qual-filter         filter reads based on minimum index quality [default: no quality filter]
    -p MIN_QUAL, --min-qual MIN_QUAL
                          reject seqs with mean barcode quality score less than this value; for dual index, both barcodes must meet the threshold [26]
    --encoding {phred}    quality score encoding; see https://en.wikipedia.org/wiki/FASTQ_format [phred]


Installation
============

Install from PyPi using pip::

  pip install barcodecop

Or install directly from the git repository::

  pip install git+https://github.com/nhoffman/barcodecop.git

Testing
=======

Run all tests like this::

  python setup.py test


CHANGES.rst
=========================
 Changelog for barcodcop
=========================

version 0.5.2-dev
=================

* `--names` file.txt[.bz2|.gz] option to output seq names that pass filtering

version 0.5.1
=============

* use GitHub CI for testing and Docker image build

version 0.5
===========

* handles empty files without error (``--allow-empty``) (GH10)
* add ``-C/--read-counts`` to provide input and output read counts
* rename ``-c/--show-counts`` to ``-b/--barcode-counts``

version 0.4.1
=============

* python3 support

version 0.4
===========

* added ``--match-filter`` and made match filtering optional

version 0.3
===========

* implement barcode quality filtering


version 0.2
===========

* publish to PyPi


tests/test.py
#!/usr/bin/env python

import sys
import os
from os import path
import inspect
from unittest import TestCase
import csv

# in python2.7, io.StringIO cannot accept str output
try:
    from cStringIO import StringIO as IO
except ImportError:
    from io import StringIO as IO

try:
    from itertools import zip_longest
except ImportError:
    from itertools import izip_longest as zip_longest

# from fastalite import fastalite, fastqlite, Opener
from barcodecop.barcodecop import main

testfiles = 'testfiles'
barcodes = path.join(testfiles, 'barcodes.fastq.gz')
barcodes_qual = path.join(testfiles, 'barcodes_qual.fastq.gz')
empty = path.join(testfiles, 'empty.fastq.gz')
outdir = 'test_output'
most_common = 'TATTACTCTA'
dual1 = path.join(testfiles, 'dual_I1.fastq.gz')
dual2 = path.join(testfiles, 'dual_I2.fastq.gz')
dual_qual1 = path.join(testfiles, 'dual_I1_qual.fastq.gz')
dual_qual2 = path.join(testfiles, 'dual_I2_qual.fastq.gz')
most_common_dual = 'ACTGGTAGGA+TTCTCTCCAG'

TEST_OUTPUT = 'test-output'

class Capturing(list):

    """
    From http://stackoverflow.com/questions/16571150/how-to-capture-stdout-output-from-a-python-function-call
    """

    def __enter__(self):
        self._stdout = sys.stdout
        sys.stdout = self._stringio = IO()
        return self

    def __exit__(self, *args):
        self.extend(self._stringio.getvalue().splitlines())
        del self._stringio   # free up some memory
        sys.stdout = self._stdout


def mkoutdir(cls, basedir=TEST_OUTPUT):
    stacknames = [x[3] for x in inspect.stack()]
    testfun = [name for name in stacknames if name.startswith('test_')][0]
    pth = path.join(basedir, cls.__class__.__name__, testfun)

    try:
        os.makedirs(pth)
    except OSError:
        pass

    return pth


def grouper(iterable, n, fillvalue=None):
    "Collect data into fixed-length chunks or blocks"
    # grouper('ABCDEFG', 3, 'x') --> ABC DEF Gxx
    args = [iter(iterable)] * n
    return zip_longest(fillvalue=fillvalue, *args)


class TestSingleIndex(TestCase):

    def test_01(self):
        with Capturing() as output:
            # because main() is called in the same global context in
            # each test, '-q' silences logging for all invocations.
            main([barcodes, '-c', '-q'])
        self.assertEqual(output[0].split('\t')[0], most_common)

    def test_02(self):
        # filter barcode file using itself; get back only the most common
        # barcode
        with Capturing() as output:
            main([barcodes, '-f', barcodes, '--match-filter'])
        desc, seqs, __, quals = list(zip(*grouper(output, 4)))
        self.assertSetEqual(set(seqs), {most_common})

    def test_03(self):
        # --head returns the specified number of records
        with Capturing() as output:
            main([barcodes, '-f', barcodes, '--head', '10'])
        desc, seqs, __, quals = list(zip(*grouper(output, 4)))
        self.assertEqual(len(seqs), 10)

    def test_04(self):
        # --invert option removes all instances of the most common bc
        with Capturing() as output:
            main([barcodes, '-f', barcodes, '--invert', '--match-filter'])
        desc, seqs, __, quals = list(zip(*grouper(output, 4)))
        self.assertNotIn(most_common, set(seqs))

    def test_05(self):
        # test warning when recovery is below --min-pct-assignment
        with Capturing() as output:
            main([barcodes, '-f', barcodes, '--min-pct-assignment', '100'])

    def test_06(self):
        # test error with --strict
        with Capturing() as output:
            self.assertRaises(
                SystemExit,
                main,
                [barcodes, '-f', barcodes, '--min-pct-assignment', '100', '--strict'])

    def test_qual_01(self):
        # test quality filtering with defaults
        with Capturing() as output:
            main([barcodes_qual, '-f', barcodes_qual, '--qual-filter'])
        desc, seqs, __, quals = list(zip(*grouper(output, 4)))
        self.assertEqual(len(seqs), 4)

    def test_qual_02(self):
        # test quality filtering with reduced threshold
        with Capturing() as output:
            main(
                [barcodes_qual, '-f', barcodes_qual, '--qual-filter', '-p', '2'])
        desc, seqs, __, quals = list(zip(*grouper(output, 4)))
        self.assertEqual(len(seqs), 5)

    def test_count(self):
        outdir = mkoutdir(self)
        read_counts = path.join(outdir, 'counts.csv')
        main([barcodes, '-f', barcodes,
              '--qual-filter',
              '-o', path.join(outdir, 'filtered.fastq'),
              '--read-counts', read_counts])

        with open(read_counts) as f:
            reader = csv.reader(f)
            fname, input, output = next(reader)
        self.assertEqual(input, '15000')
        self.assertEqual(output, '14729')

    def test_empty(self):
        # test filtering of empty fastq
        # with Capturing() as output:

        outdir = mkoutdir(self)
        read_counts = path.join(outdir, 'counts.csv')
        main([empty, '-f', empty,
              '-o', path.join(outdir, 'filtered.fastq'),
              '--read-counts', read_counts])

        with open(read_counts) as f:
            reader = csv.reader(f)
            fname, input, output = next(reader)
        self.assertEqual(input, '0')
        self.assertEqual(output, '0')

    # TODO: test non-default offset, but will need sequences with the
    # appropriate encoding

    # def test_qual_03(self):
    #     # test quality filtering with modified offset
    #     with Capturing() as output:
    #         main(
    #             [barcodes_qual,
    #              '-f',
    #              barcodes_qual,
    #              '--qual-filter',
    #              '-p',
    #              '1',
    #              '--qual-offset',
    #              '34'])
    #     desc, seqs, __, quals = zip(*grouper(output, 4))
    #     self.assertEqual(len(seqs), 5)


class TestDualIndex(TestCase):

    def test_01(self):
        with Capturing() as output:
            # because main() is called in the same global context in
            # each test, '-q' silences logging for all invocations.
            main([dual1, dual2, '-c'])
        self.assertEqual(output[0].split('\t')[0], most_common_dual)

    def test_02(self):
        # filter barcode file using itself; get back only the most common
        # barcode
        with Capturing() as output:
            main([dual1, dual2, '-f', dual1, '--match-filter'])
        desc, seqs, __, quals = list(zip(*grouper(output, 4)))
        self.assertSetEqual(set(seqs), {most_common_dual.split('+')[0]})

    def test_qual_dual(self):
        # test quality filtering with defaults
        with Capturing() as output:
            main([dual_qual1, dual_qual2, '-f', dual_qual1, '--qual-filter'])
        desc, seqs, __, quals = list(zip(*grouper(output, 4)))
        self.assertEqual(len(seqs), 5)


barcodecop/barcodecop.py
#!/usr/bin/env python

"""Filter fastq files by enforcing exact barcode match and quality score.

Input and output files may be compressed as indicated by a .bz2 or .gz
suffix.

"""

import argparse
import sys
import csv
from collections import Counter
import logging
from collections import namedtuple

from itertools import islice, tee
try:
    from itertools import filterfalse, zip_longest
except ImportError:
    from itertools import izip_longest as zip_longest
    from itertools import ifilterfalse as filterfalse
    from itertools import ifilter as filter


from fastalite import fastqlite, Opener

try:
    from . import __version__
except:
    __version__ = ''

# default values
MIN_QUAL = 26


class VersionAction(argparse._VersionAction):

    """Write the version string to stdout (rather than stderr) and exit"""

    def __call__(self, parser, namespace, values, option_string=None):
        formatter = parser._get_formatter()
        formatter.add_text(
            parser.version if self.version is None else self.version)
        sys.stdout.write(formatter.format_help())
        sys.exit(0)


def get_match_filter(barcode):
    """Return a function for filtering a pair of (seq, bc) namedtuple
    pairs; the function returns True if bc.seq == barcode

    """

    def filterfun(pair):
        seq, bc = pair
        return str(bc.seq) == barcode

    return filterfun


def get_qual_filter(min_qual, encoding, paired=False):
    """Return a function for filtering a pair of (seq, bc) namedtuple
    pairs. The function returns True if the average barcode quality
    score calculated using the provided encoding method is at least
    min_qual. If``paired`` is True, ``bc`` must be a namedtuple with
    attributes qual and qual2; the function returns True if the average
    barcode quality score calculated using the provided encoding method
    is at least min_qual.  The function defined for each encoding method
    is specified as ``get_{}_encoding``.  Currently only Sanger phred
    encoding is supported -- see ``get_phred_encoding``.

    """

    encoding = globals()['get_{}_encoding'.format(encoding)]()

    if paired:
        def filterfun(pair):
            seq, bc = pair
            return (check_score(encoding, min_qual, bc.qual) and
                    check_score(encoding, min_qual, bc.qual2))
    else:
        def filterfun(pair):
            seq, bc = pair
            return check_score(encoding, min_qual, bc.qual)

    return filterfun


def get_phred_encoding():
    """Return a dict of {ascii character: quality score} providing an
    encoding of ASCII characters 33 to 126 corresponding to values 0
    to 93.

    see https://en.wikipedia.org/wiki/FASTQ_format

    """

    offset = 33
    return {chr(i): i - offset for i in range(offset, 127)}


def check_score(encoding, min_qual, qual_str):
    """Return True if the average quality score is at least min_qual

    """
    qscores = [encoding[q] for q in qual_str]
    return sum(qscores) >= min_qual * len(qscores)


def seqdiff(s1, s2):
    if s1 == s2:
        return s1
    else:
        return ''.join('.' if c1 == c2 and c1.isalpha() else c2
                       for c1, c2 in zip(s1, s2))


def as_fastq(seq):
    return u'@{seq.description}\n{seq.seq}\n+\n{seq.qual}\n'.format(seq=seq)


def combine_dual_indices(file1, file2):
    Seq = namedtuple('Seq', ['id', 'seq', 'qual', 'qual2'])
    for i1, i2 in zip_longest(fastqlite(file1), fastqlite(file2)):
        assert i1.id == i2.id
        yield Seq(id=i1.id, seq=i1.seq + '+' + i2.seq, qual=i1.qual, qual2=i2.qual)


def close_all_files(args):
    for name, obj in args.__dict__.items():
        if (obj and hasattr(obj, 'close') and
            hasattr(obj, 'closed') and hasattr(obj, 'name')):
            if not obj.closed:
                obj.close()


def main(arguments=None):
    parser = argparse.ArgumentParser(
        prog='barcodecop', description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument(
        'index', nargs='+', type=Opener(), metavar='file.fastq[.bz2|.gz]',
        help='one or two files containing index reads in fastq format')
    parser.add_argument(
        '-o', '--outfile', default=sys.stdout, type=Opener('w'),
        help='output fastq')
    parser.add_argument(
        '--snifflimit', type=int, default=10000, metavar='N',
        help='read no more than N records from the index file [%(default)s]')
    parser.add_argument(
        '--head', type=int, metavar='N',
        help='limit the output file to N records')
    parser.add_argument(
        '--invert', action='store_true', default=False,
        help='include only sequences failing filtering criteria')
    parser.add_argument(
        '-q', '--quiet', action='store_true', default=False,
        help='minimize messages to stderr')
    parser.add_argument(
        '-V', '--version', action=VersionAction, version=__version__,
        help='Print the version number and exit')

    processing_options = parser.add_mutually_exclusive_group(required=True)
    processing_options.add_argument(
        '-c', '--show-counts', action='store_true', default=False,
        help='tabulate barcode counts and exit')
    processing_options.add_argument(
        '-f', '--fastq', type=Opener(), metavar='file.fastq[.bz2|.gz]',
        help='reads to filter in fastq format')

    match_options = parser.add_argument_group('Barcode matching options')

    match_options.add_argument(
        '--match-filter', action='store_true', default=False,
        help=('filter reads based on exact match to most common barcode '
              '[default: no match filter]'))
    match_options.add_argument(
        '--min-pct-assignment', type=float, default=90.0, metavar='PERCENT',
        help=("""warn (or fail with an error; see --strict) if the
               most common barcode represents less than PERCENT of the
               total [%(default)s]"""))
    match_options.add_argument(
        '--strict', action='store_true', default=False,
        help=("""fail if conditions of --min-pct-assignment are not met"""))
    match_options.add_argument(
        '-b', '--barcode-counts', type=argparse.FileType('w'), metavar='FILE',
        help='tabulate barcode counts and store as a CSV')
    match_options.add_argument(
        '-C', '--read-counts', type=argparse.FileType('w'), metavar='FILE',
        help='tabulate read counts and store as a CSV')
    match_options.add_argument(
        '--allow-empty', action='store_true', default=False,
        help=('accept fastq files which contain zero length seqs/qual strings'
              '[default: do not allow empty seqs in fastq files]'))

    qual_options = parser.add_argument_group('Barcode quality filtering options')

    qual_options.add_argument(
        '--qual-filter', action='store_true', default=False,
        help=('filter reads based on minimum index quality '
              '[default: no quality filter]'))
    qual_options.add_argument(
        '-p', '--min-qual', type=int, default=MIN_QUAL,
        help="""reject seqs with mean barcode quality score less than
        this value; for dual index, both barcodes must meet the
        threshold [%(default)s]""")
    qual_options.add_argument(
        '--encoding', default='phred', choices=['phred'],
        help="""quality score encoding; see
             https://en.wikipedia.org/wiki/FASTQ_format [%(default)s]""")

    args = parser.parse_args(arguments)

    logging.basicConfig(
        format='%(message)s',
        level=logging.ERROR if args.quiet else logging.WARNING)
    log = logging.getLogger(__name__)

    # when provided with dual barcodes, concatenate into a single
    # namedtuple with attributes qual and qual1; generate a filter
    # function appropriate for either case.
    if len(args.index) == 1:
        bcseqs = fastqlite(args.index[0])
        qual_filter = get_qual_filter(args.min_qual, args.encoding)
    elif len(args.index) == 2:
        qual_filter = get_qual_filter(args.min_qual, args.encoding, paired=True)
        bcseqs = combine_dual_indices(*args.index)
    else:
        log.error('error: please specify either one or two index files')

    # use bc1 to determine most common barcode
    bc1, bc2 = tee(bcseqs, 2)

    iseqs = [str(seq.seq) for seq in islice(bc1, args.snifflimit)]

    # determine the most common barcode
    if iseqs:
        barcode_counts = Counter(iseqs)
    else:
        barcode_counts = Counter({None: 0})

    barcodes, counts = list(zip(*barcode_counts.most_common()))

    most_common_bc = barcodes[0]
    try:
        most_common_pct = 100 * float(counts[0]) / sum(counts)
    except ZeroDivisionError:
        most_common_pct = 0
    else:
        if most_common_pct < args.min_pct_assignment:
            msg = 'frequency of most common barcode is less than {}%'.format(
                args.min_pct_assignment)
            if args.strict:
                log.error('Error: ' + msg)
                sys.exit(1)
            else:
                log.warning('Warning: ' + msg)

    log.info('most common barcode: {} ({}/{} = {:.2f}%)'.format(
        most_common_bc, counts[0], sum(counts), most_common_pct))

    if args.barcode_counts:
        # Create a writer using the CSV module
        barcode_counts_writer = csv.writer(args.barcode_counts)
        # Write a header
        barcode_counts_writer.writerow(['barcode', 'diff_most_common', 'count'])
        for bc, count in barcode_counts.most_common():
            barcode_counts_writer.writerow([bc, seqdiff(most_common_bc, bc), count])

    if args.show_counts:
        for bc, count in barcode_counts.most_common():
            print(('{}\t{}\t{}'.format(bc, seqdiff(most_common_bc, bc), count)))
        return None

    ifilterfun = filterfalse if args.invert else filter

    # use seqs2 for --read-counts input-count
    seqs, seqs2 = tee(fastqlite(args.fastq, args.allow_empty))

    filtered = zip_longest(seqs, bc2)

    if args.match_filter:
        filtered = ifilterfun(get_match_filter(most_common_bc), filtered)

    if args.qual_filter:
        filtered = ifilterfun(qual_filter, filtered)

    if args.read_counts:
        filtered, filtered2 = tee(filtered)
        input_count = sum(1 for _ in seqs2)
        filtered_count = sum(1 for _ in filtered2)
        if args.head is not None:
            output_count = min(filtered_count, args.head)
        else:
            output_count = filtered_count
        read_counts_writer = csv.writer(args.read_counts)
        read_counts_writer.writerow(
            ['outfile', 'in', 'filtered', 'out'])
        read_counts_writer.writerow(
            [args.outfile.name, input_count, filtered_count, output_count])

    for seq, bc in islice(filtered, args.head):
        assert seq.id == bc.id
        args.outfile.write(as_fastq(seq))

    close_all_files(args)


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
